### PR TITLE
[FEATURE] Add frontend linting

### PR DESCRIPTION
**Pull request by lukaszuznanski**
Add eslint with recommended config for typescript
Add styleslint with recommended config

### CODE REVIEW

**Comment by lukaszuznanski** [comment]
@oliverklee configuration files moved to Private/Resources, sample files are in Resources/Public, added also gitattributes to remove config files from build.

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Resources/Private/.eslintignore export-ignore
+Resources/Private/.eslintrc export-ignore
+Resources/Private/.prettierrc.js export-ignore
+Resources/Private/package.json export-ignore
+Resources/Private/stylelint.config.js export-ignore
```
Comment (oliverklee) [inline .gitattributes:5]: Please let's have a LF after the last line so adding a new line does not modify the not-the-last-anymore line. :-)

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -57,6 +57,26 @@ jobs:
...
+        uses: actions/checkout@v2
+      -
+        name: "Install modules"
+        working-directory: ./Resources/Private
+        run: yarn
+      -
+        name: "Run command"
+        working-directory: ./Resources/Private
+        run: "yarn lint:${{ matrix.command }}"
+    strategy:
```
Comment (oliverklee) [inline .github/workflows/ci.yml:75]: Please let's move this up so the reader already has read the matrix when the matrix variable is used.

File: .gitignore [hunk 3]
```diff
@@ -4,3 +4,7 @@
 /composer.lock
 /nbproject
 /var
+/Resources/Private/node_modules/
+/Resources/Private/yarn.lock
+/Resources/Private/yarn-error.log
+/Resources/Private/package-lock.json
```
Comment (oliverklee) [inline .gitignore:10]: Let's have a LF after the last line, please.

File: Resources/Private/.eslintignore [hunk 4]
```diff
@@ -0,0 +1 @@
+node_modules
```
Comment (oliverklee) [inline Resources/Private/.eslintignore:1]: LF at the end, please.

File: Resources/Private/.eslintrc.json [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    "@typescript-eslint"
+  ],
+  "extends": [
+    "eslint:recommended",
+    "plugin:@typescript-eslint/eslint-recommended",
+    "plugin:@typescript-eslint/recommended",
+    "prettier/@typescript-eslint",
+    "plugin:prettier/recommended"
+  ]
+}
```
Comment (oliverklee) [inline Resources/Private/.eslintrc.json:21]: LF at the end, please.

File: Resources/Private/.prettierrc.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+    semi: true,
+    trailingComma: "all",
+    singleQuote: true,
+    printWidth: 120,
+    tabWidth: 4
+};
```
Comment (oliverklee) [inline Resources/Private/.prettierrc.js:7]: LF at the end, please.

File: Resources/Private/package.json [hunk 7]
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "example-extension",
```
Comment (oliverklee) [inline Resources/Private/package.json:2]: Can we have a more descriptive name here?

File: Resources/Private/package.json [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  },
+  "keywords": [],
+  "author": "",
+  "version": "1.0.0",
+  "license": "ISC",
+  "scripts": {
+    "lint:js": "eslint --config .eslintrc.json '../Public/*/**/*.{js,ts,tsx}'",
+    "lint:js:fix": "eslint --config .eslintrc.json '../Public/*/**/*.{js,ts,tsx}' --quiet --fix",
+    "lint:style": "stylelint --config stylelint.config.js ../Public/*/*.{css,scss,less}",
+    "lint:style:fix": "stylelint --config stylelint.config.js ../Public/*/*.{css,scss,less} --fix"
```
Comment (oliverklee) [inline Resources/Private/package.json:15]: Are 2 spaces (not 4) for JSON current best practice?

File: Resources/Private/stylelint.config.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        "selector-pseudo-element-colon-notation": "double",
+        "selector-type-case": "lower",
+        "unit-case": "lower",
+        "value-keyword-case": "lower",
+        "value-list-comma-newline-after": "always-multi-line",
+        "value-list-comma-space-after": "always-single-line",
+        "value-list-comma-space-before": "never",
+        "value-list-max-empty-lines": 0,
+    },
+};
```
Comment (oliverklee) [inline Resources/Private/stylelint.config.js:110]: LF at the end, please.

File: Resources/Public/Css/Main.css [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+.selector-a,
+.selector-b:not(:first-child) {
+  padding: 10px !important;
+  top: calc(calc(1em * 2) / 3);
+}
+
+.selector-x { width: 10%; }
+.selector-y { width: 20%; }
+.selector-z { width: 30%; }
```
Comment (oliverklee) [inline Resources/Public/Css/Main.css:25]: Would it make sense to have some real CSS here, e.g., for the tea list items?

File: Resources/Public/Css/Main.css [hunk 11]
```diff
@@ -0,0 +1,69 @@
+@import url(x.css);
```
Comment (oliverklee) [inline Resources/Public/Css/Main.css:1]: As we have real files in `Resources/Public/CSS/`, the `Resources/Public/CSS/.gitkeep` is not needed anymore.

File: Resources/Public/JavaScript/Test.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
+const menuConfig = {
```
Comment (oliverklee) [inline Resources/Public/JavaScript/Test.js:1]: As we have real files in `Resources/Public/JavaScript/`, the `Resources/Public/JavaScript.gitkeep` is not needed anymore.

File: Resources/Public/Scss/Main.scss [hunk 13]
```diff
@@ -0,0 +1,9 @@
+$bgcolor: lightblue;
+$textcolor: darkblue;
+$fontsize: 18px;
```
Comment (oliverklee) [inline Resources/Public/Scss/Main.scss:3]: Do we currently compile this file?

File: Resources/Public/TypeScript/Test.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
+class Student {
```
Comment (oliverklee) [inline Resources/Public/TypeScript/Test.ts:1]: Do we currently compile this file?

File: Resources/Public/JavaScript/Test.js [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+  title: "Order",
+  buttonText: "Send",
+  cancellable: true,
+};
+
+function createMenu(config) {
```
Comment (oliverklee) [inline Resources/Public/JavaScript/Test.js:7]: Can we have some JS here that does something real for the Tea extension?

File: Resources/Public/Css/Main.css (hunk 11 above)
Comment (lukaszuznanski) [reply Resources/Public/Css/Main.css:1]: it's removed already

File: Resources/Public/Css/Main.css (hunk 10 above)
Comment (lukaszuznanski) [reply Resources/Public/Css/Main.css:25]: ok, fine :)

File: Resources/Public/JavaScript/Test.js (hunk 12 above)
Comment (lukaszuznanski) [reply Resources/Public/JavaScript/Test.js:1]: yea it's removed already

File: Resources/Public/JavaScript/Test.js (hunk 15 above)
Comment (lukaszuznanski) [reply Resources/Public/JavaScript/Test.js:7]: ok :)

File: Resources/Public/TypeScript/Test.ts (hunk 14 above)
Comment (lukaszuznanski) [reply Resources/Public/TypeScript/Test.ts:1]: no, but I can add compiling

**Comment by oliverklee** [comment]
I'll review this later today (after our call).

In the meantime, could your please rebase (just to make sure we have the latest changes here)?

**Comment by lukaszuznanski** [comment]
remove typescript, remove sass, js disable submit button

**Review by oliverklee — CHANGES_REQUESTED** [review]
Needs some changes as discussed in the call (as as written done in your latest comment).

**Review by oliverklee — CHANGES_REQUESTED** [review]
The JS and CSS checks currently are failing as they do not find any files. Strange.

**Comment by lukaszuznanski** [comment]
Yes, I removed them and as I said on status call - I need add new :).

**Comment by lukaszuznanski** [comment]
@oliverklee 
we have css and js file which is sorting table. Should we minify those files at the end?

**Comment by oliverklee** [comment]
> we have css and js file which is sorting table. Should we minify those files at the end?

@lukaszuznanski No, I don't think so (at least not within the extensions). This is the site's responsibility.

File: Resources/Private/.htaccess [hunk 16]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (oliverklee) [inline Resources/Private/.htaccess:1]: Why this change?

File: Resources/Private/.eslintrc.json [hunk 17]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (oliverklee) [inline Resources/Private/.eslintrc.json:1]: This needs to be in the `.gitatttributes`. (Actually, I think the entry there is incorrect: `/Resources/Private/.eslintrc`)

File: Resources/Private/Templates/Tea/Index.html [hunk 18]
```diff
@@ -8,6 +8,12 @@ <h2>
         </h2>
 
         <table class="table">
+            <thead>
+                <tr>
+                    <th>a</th>
+                    <th>b</th>
```
Comment (oliverklee) [inline Resources/Private/Templates/Tea/Index.html:14]: Please let's have meaningful headings here. I propose the localized labels "UID" and "Title".

File: Resources/Private/package.json [hunk 19]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "tea",
+  "description": "",
+  "repository": {
+    "url": "https://github.com/TYPO3-Documentation/tea.git"
+  },
+  "keywords": [],
+  "author": "",
+  "version": "1.0.0",
+  "license": "ISC",
```
Comment (oliverklee) [inline Resources/Private/package.json:10]: Shouldn't this be the same license as the rest of the extension, i.e., GPL V2+?

File: Resources/Public/JavaScript/Main.js [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+        (function (i) {
+            var dir = 1;
+            th[i].addEventListener("click", function () {
+                sortTable(table, i, (dir = 1 - dir));
+            });
+        })(i);
+}
+
+function sortTable(table, col, reverse) {
+    var tb = table.tBodies[0], //
```
Comment (oliverklee) [inline Resources/Public/JavaScript/Main.js:23]: Why the `//` here?

File: Resources/Public/JavaScript/Main.js [hunk 21]
```diff
@@ -0,0 +1,36 @@
+document.addEventListener("DOMContentLoaded", function () {
+    var t = document.querySelectorAll(".tx-tea table"),
+        i = t.length;
+    while (--i >= 0) makeSortable(t[i]);
+});
+
+function makeSortable(table) {
```
Comment (oliverklee) [inline Resources/Public/JavaScript/Main.js:7]: Should we namespace the functions to avoid name collisions, e.g., like this?
https://github.com/oliverklee/ext-seminars/blob/master/Resources/Public/JavaScript/FrontEnd/FrontEnd.js#L22

File: Resources/Public/JavaScript/Main.js [hunk 22]
```diff
@@ -0,0 +1,36 @@
+document.addEventListener("DOMContentLoaded", function () {
+    var t = document.querySelectorAll(".tx-tea table"),
+        i = t.length;
+    while (--i >= 0) makeSortable(t[i]);
```
Comment (oliverklee) [inline Resources/Public/JavaScript/Main.js:4]: Please let's have the `{` and `}` here. (Shouldn't eslint have complained about this?)

**Review by oliverklee — CHANGES_REQUESTED** [review]
